### ISSUE TITLE
transport tunnel using IKEv2 and %any as right end does not work

### ISSUE THREAD

**Issue by nekonyuu** [xelerance/Openswan#122]
Hello,

With the following configuration, I cannot mount the tunnel between two machines with IKEv2 set to "propose". Disabling IKEv2 altogether allows me to mount the tunnel correctly.
## Initiating side of the tunnel

```
conn client
   authby=secret
   type=transport
   right=172.31.36.127
   rightprotoport=tcp/1718
   left=%eth0
   leftprotoport=tcp/0
   auto=start
   ike=aes256-sha1;modp2048
   ikev2=propose
   phase2=esp
   phase2alg=aes256-sha1;modp2048
   pfs=yes
   salifetime=86400s
   ikelifetime=86400s
   dpddelay=2
   dpdtimeout=10
   dpdaction=restart
```
## Server side of the tunnel

```
conn server
   authby=secret
   type=transport
   right=%any
   rightprotoport=tcp/0
   left=%eth0
   leftprotoport=tcp/1718
   auto=add
   ike=aes256-sha1;modp2048
   ikev2=permit
   phase2=esp
   phase2alg=aes256-sha1;modp2048
   pfs=yes
   salifetime=86400s
   ikelifetime=86400s
   dpddelay=2
   dpdtimeout=10
   dpdaction=restart
```
## Relevant logs (plutodebug=all)

```
Apr  8 12:41:59 scoring-1 pluto[1987]: | processing payload: ISAKMP_NEXT_v2Ni (len=20) 
Apr  8 12:41:59 scoring-1 pluto[1987]: | Now lets proceed with payload (ISAKMP_NEXT_v2V)
Apr  8 12:41:59 scoring-1 pluto[1987]: | ***parse IKEv2 Vendor ID Payload:
Apr  8 12:41:59 scoring-1 pluto[1987]: |    next payload type: ISAKMP_NEXT_NONE
Apr  8 12:41:59 scoring-1 pluto[1987]: |    critical bit: none
Apr  8 12:41:59 scoring-1 pluto[1987]: |    length: 16
Apr  8 12:41:59 scoring-1 pluto[1987]: | processing payload: ISAKMP_NEXT_v2V (len=16) 
Apr  8 12:41:59 scoring-1 pluto[1987]: | Finished and now at the end of ikev2_process_payload
Apr  8 12:41:59 scoring-1 pluto[1987]: | Finished processing ikev2_process_payloads
Apr  8 12:41:59 scoring-1 pluto[1987]: | Now lets proceed with state specific processing
Apr  8 12:41:59 scoring-1 pluto[1987]: | find_host_connection2 called from ikev2parent_inI1outR1, me=10.10.0.20:500 him=10.10.0.238:500 policy=IKEv2ALLOW
Apr  8 12:41:59 scoring-1 pluto[1987]: | find_host_pair: comparing to 10.10.0.20:500 10.10.0.10:500 
Apr  8 12:41:59 scoring-1 pluto[1987]: | find_host_pair: comparing to 10.10.0.20:500 10.10.1.10:500 
Apr  8 12:41:59 scoring-1 pluto[1987]: | find_host_pair: comparing to 10.10.0.20:500 10.10.2.10:500 
Apr  8 12:41:59 scoring-1 pluto[1987]: | find_host_pair: comparing to 10.10.0.20:500 10.10.0.4:500 
Apr  8 12:41:59 scoring-1 pluto[1987]: | find_host_pair: comparing to 10.10.0.20:500 0.0.0.0:500 
Apr  8 12:41:59 scoring-1 pluto[1987]: | find_host_pair_conn (find_host_connection2): 10.10.0.20:500 10.10.0.238:500 -> hp:none 
Apr  8 12:41:59 scoring-1 pluto[1987]: | searching for connection with policy = IKEv2ALLOW
Apr  8 12:41:59 scoring-1 pluto[1987]: | find_host_connection2 returns empty
Apr  8 12:41:59 scoring-1 pluto[1987]: | no connection found 
Apr  8 12:41:59 scoring-1 pluto[1987]: | complete v2 state transition with STF_FAIL
Apr  8 12:41:59 scoring-1 pluto[1987]: packet from 10.10.0.238:500: sending  notification v2N_NO_PROPOSAL_CHOSEN to 10.10.0.238:500
Apr  8 12:41:59 scoring-1 pluto[1987]: | **emit ISAKMP Message:
Apr  8 12:41:59 scoring-1 pluto[1987]: |    initiator cookie:
Apr  8 12:41:59 scoring-1 pluto[1987]: |   d2 1d c1 cd  1a b4 a9 6d
Apr  8 12:41:59 scoring-1 pluto[1987]: |    responder cookie:
Apr  8 12:41:59 scoring-1 pluto[1987]: |   00 00 00 00  00 00 00 00
Apr  8 12:41:59 scoring-1 pluto[1987]: |    next payload type: ISAKMP_NEXT_v2N
Apr  8 12:41:59 scoring-1 pluto[1987]: |    ISAKMP version: IKEv2 version 2.0 (rfc4306/rfc5996)
Apr  8 12:41:59 scoring-1 pluto[1987]: |    exchange type: ISAKMP_v2_SA_INIT
Apr  8 12:41:59 scoring-1 pluto[1987]: |    flags: ISAKMP_FLAG_RESPONSE
Apr  8 12:41:59 scoring-1 pluto[1987]: |    message ID:  00 00 00 00
Apr  8 12:41:59 scoring-1 pluto[1987]: | Adding a v2N Payload
Apr  8 12:41:59 scoring-1 pluto[1987]: | ***emit IKEv2 Notify Payload:
Apr  8 12:41:59 scoring-1 pluto[1987]: |    next payload type: ISAKMP_NEXT_NONE
Apr  8 12:41:59 scoring-1 pluto[1987]: |    critical bit: none
Apr  8 12:41:59 scoring-1 pluto[1987]: |    Protocol ID: PROTO_ISAKMP
Apr  8 12:41:59 scoring-1 pluto[1987]: |    SPI size: 0
Apr  8 12:41:59 scoring-1 pluto[1987]: |    Notify Message Type: v2N_NO_PROPOSAL_CHOSEN
Apr  8 12:41:59 scoring-1 pluto[1987]: | emitting length of IKEv2 Notify Payload: 8
Apr  8 12:41:59 scoring-1 pluto[1987]: | emitting length of ISAKMP Message: 36
Apr  8 12:41:59 scoring-1 pluto[1987]: | sending 36 bytes for send_v2_notification through eth0:500 to 10.10.0.238:500 (using #0)
Apr  8 12:41:59 scoring-1 pluto[1987]: |   d2 1d c1 cd  1a b4 a9 6d  00 00 00 00  00 00 00 00
Apr  8 12:41:59 scoring-1 pluto[1987]: |   29 20 22 20  00 00 00 00  00 00 00 24  00 00 00 08
Apr  8 12:41:59 scoring-1 pluto[1987]: |   01 00 00 0e
Apr  8 12:41:59 scoring-1 pluto[1987]: | state transition function for no-state failed: NO_PROPOSAL_CHOSEN
```

It seems that on the server side, the %any wildcard on right is not supported/detected as such, rendering any proposal using this unusable for openswan.

I'm currently testing a potential fix for this, as the IKEv2 part of pluto does not seems to use the same functions for host pair matching.

**Comment by letoams** [comment]
On Wed, 8 Apr 2015, Jonathan Raffre wrote:

> With the following configuration, I cannot mount the tunnel between two machines with IKEv2 set to "propose". Disabling IKEv2 altogether
> allows me to mount the tunnel correctly.

openswan has some known bugs with ikev2 and transport mode. Those should
be addressed in the rhel6/centos6 package of openswan though. Libreswan
also works fine for this.

>    ikev2=propose

>    ikev2=permit

Note this combination could yield a ikev1 tunnel of the "permit" side
would initiate.

You might want ikev2=insist on both ends.

Have you tried not using %eth0 but %defaultroute instead?

> Apr  8 12:41:59 scoring-1 pluto[1987]: | find_host_pair_conn (find_host_connection2): 10.10.0.20:500 10.10.0.238:500 -> hp:none 
> Apr  8 12:41:59 scoring-1 pluto[1987]: | searching for connection with policy = IKEv2ALLOW
> Apr  8 12:41:59 scoring-1 pluto[1987]: | find_host_connection2 returns empty
> Apr  8 12:41:59 scoring-1 pluto[1987]: | no connection found 
> Apr  8 12:41:59 scoring-1 pluto[1987]: | complete v2 state transition with STF_FAIL

This suggests no connection was allowed with IKEv2. see "ipsec auto --status" on scoring-1 and look at the policy
flags for that connection?

> It seems that on the server side, the %any wildcard on right is not supported/detected as such, rendering any proposal using this
> unusable for openswan.

It should work, but the last few versions of openswan have not been
tested by me. I know some regression has slipped in, which is why
RHEL6 is not rebasing to a newer version (and RHEL7 obsoleted openswan
in favour of libreswan)

> I'm currently testing a potential fix for this, as the IKEv2 part of pluto does not seems to use the same functions for host pair
> matching.

You might want to check the libreswan source code to compare.

Paul

**Comment by nekonyuu** [comment]
I'm working with the source distirbution for openswan (on ArchLinux) and tested on Ubuntu 14.04 with the same behavior. As you suggested I tested with ikev2=insist on both end with the same result.

I'll test with CentOS 6 and libreswan though, but the part of the code managing the hostpair checking is the same between the two codebase when I checked before making this issue so I supposed libreswan would have the same bug, from source at least.

**Comment by nekonyuu** [comment]
Just tested with CentOS 6 with the same behavior, tried with ikev2=insist and using %defaultroute instead of %eth0.

```
[root@i-fd48031b ~]$ rpm -q openswan
openswan-2.6.32-37.el6.x86_64
[root@i-fd48031b ~]$ cat /etc/redhat-release 
CentOS release 6.6 (Final)
```

The following fix nekonyuu/Openswan@6f9b3cbe25cce44467967abda0f5a8c7a7516cab seems to fix this for me, although i dunno if it would be the right way for doing this for now and it is not sufficiently tested with others configurations on my side.

**Comment by nekonyuu** [comment]
Just tested on CentOS 7 with libreswan, with success !

Debug logs suggested libreswan was doing a double check somewhere with %any when the first search was returning NULL, after some digging I found that libreswan does effectively this there : https://github.com/libreswan/libreswan/blob/master/programs/pluto/ikev2_parent.c#L657-L718 .

Could it be possible for openswan to backport this ?

**Comment by letoams** [comment]
On Thu, 9 Apr 2015, Jonathan Raffre wrote:

> Just tested on CentOS 7 with libreswan, with success !

Great :)

> Debug logs suggested libreswan was doing a double check somewhere with %any when the first search was returning NULL, after some
> digging I found that libreswan does effectively this there :
> https://github.com/libreswan/libreswan/blob/master/programs/pluto/ikev2_parent.c#L657-L718 .
> 
> Could it be possible for openswan to backport this ?

Oh I see. They haven't yet added support for roadwarriors in IKEv2.

Well, they have a lot of backporting to do:

https://nohats.ca/wordpress/wp-content/uploads/2014/02/animated.gif

Paul

**Comment by mcr** [comment]
Yes, there is an issue with IKEv2 and %any, and I've actually been working on a test case and fix, which I expect to release next week.

**Comment by nekonyuu** [comment]
Oh, thanks for the info. 

If you need someone who can validate it in situation, i can help :)

**Comment by shussain** [comment]
This appears to have been fixed in 2.6.47
